getPrevalence, NA values

TuomasBorman opened the pull request:
If there are NA values, they are interpreted as FALSE i.e., certain value do not exceed the detection threshold i.e. "this taxon cannot be found from this sample"

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'master' into getprevalence

TuomasBorman committed:
up

antagomir reviewed (commented):
This is good if we assume that users will always want to consider NA as "not detected". I think this is OK.

Another option would be to let users pass na.rm=TRUE for rowSums on line 216 (or perhaps this could be the default then; albeit now na.rm=FALSE by default in other parts of this R file).

It might be advisable to throw a warning when NAs are interpreted as "not detected". User should get to know this.

I would also add some unit tests to spot this problem in the future if anything is changed?

TuomasBorman commented:
One option is to catch `na.rm` in `.agg_for_prevalence` function and create new parameter called `drop.empty.rank` --> don't feed it to `agglomreateByRank` function instead use `drop.empty.rank`


--> Use na.rm in rowSums in line 216

antagomir commented:
Yes the `drop.empty.rank` might be more clear.

antagomir commented:
TBH I am not sure how necessary it is to allow users decide whether they want to have to have `na.rm` TRUE of FALSE in the merging. But perhaps it is good to let user decide ultimately. 

I would pay some attention for documentation, examples & unit tests to keep it clear for users.

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman committed:
up

TuomasBorman commented:
Is it now good to go?

antagomir commented on R/getPrevalence.R:
```diff
@@ -34,12 +34,18 @@
 #'     \code{\link[=agglomerate-methods]{?agglomerateByRank}} for more details.
+#'     Note that you can specify whether to remove empty ranks with
+#'     \code{drop.empty.ranks} instead of \code{na.rm}. (default:
+#'     \code{drop.empty.rank = FALSE})
```
should here be rank -> ranks?

antagomir replied:
or just "default: FALSE"

TuomasBorman replied:
The documentation is not unified; there are multiple forms in this file alone --> (default: something=FALSE), (default: FALSE)

--> FALSE

antagomir commented on R/getPrevalence.R:
```diff
@@ -204,13 +213,13 @@ setMethod("getPrevalence", signature = c(x = "ANY"),
         }
-        prev <- rowSums(prev)
+        # Calculate prevalence for each taxa
+        prev <- rowSums(prev, na.rm = na.rm)
```
I am still just wondering if we should throw a warning in a case where NAs actually exist & actively ignored? (ignoring NAs is now the default, user might not notice that this happens unless they are warned)

TuomasBorman replied:
That is good and adds transparency

antagomir commented on tests/testthat/test-5prevalence.R:
```diff
@@ -62,6 +62,36 @@ test_that("getPrevalence", {
+    assay(tse, "counts")[remove, ] <- NA
+    # There should be 3 NA values if na.rm = FALSE. Otherwise there should be 0
+    expect_true( sum(is.na(
+        getPrevalence(tse, assay.type = "counts", na.rm = FALSE) )) == 3)
```
does this also work when `rank` argument is specified? Because earlier there were differences how this works when rank is additionally provided, compared to this present example.

TuomasBorman replied:
These modifications do not affect this thing, I guess. (I am not sure what differences there were)

When rank is specified

1. TreeSE is agglomerated to level specified by `rank`. `agg.na.rm` (I renamed `drop.empty.ranks`) is used to specify whether to drop those rows that do not have taxonomy in level specified by `rank.` --> Instead of "phylum_something" & "class_something2" there are only taxa that have taxonomy info in class level, for instance.

2. If there are NA values, agglomerated tables have also NA values in higher level rank --> This might be misleading? We could add parameter to agglomerateByRank that removes NAs? (agglomerateByRanks --> mergeRows --> .merge_rows --> scuttle::sumCountsAcrossFeatures)

3. getPrevalence calculates prevalence of features. With `na.rm,` user can specify whether to remove NA values.


So currently, when there are NA values and rank is specified, there are some taxa that have prevalence NA.

antagomir reviewed (commented):
Some suggestions in the comments.

TuomasBorman committed:
up

TuomasBorman committed:
up

antagomir commented on R/getPrevalence.R:
```diff
@@ -213,6 +212,14 @@ setMethod("getPrevalence", signature = c(x = "ANY"), function(
+        if( any( is.na(x) ) ){
+            msg <- paste0(
+                "The abundance table contains NA values and they are ",
+                ifelse(na.rm, "not", ""), "excluded (see 'na.rm').")
```
is this in sync with the agg.na.rm?

TuomasBorman replied:
This is `na.rm` for `getPrevalence` function (`agg.na.rm` is for `agglomerateByRanks`) --> so yes

antagomir replied:
Ok. Just ensure it is sufficiently clear from documentation what na.rm is for what purposes and there is no real risk for confusion between those.

antagomir reviewed (commented):
Can we also have in unit tests test of the rank parameter, something like

getPrevalence(tse, assay.type = "counts", na.rm = FALSE, rank="Genus")

TuomasBorman committed:
up

TuomasBorman committed:
Merge branch 'master' into getprevalence

TuomasBorman committed:
Merge branch 'master' into getprevalence

TuomasBorman committed:
up

TuomasBorman committed:
up

Merge branch 'getprevalence' of github.com:microbiome/mia into getprevalence

# Conflicts:
#	DESCRIPTION
